Refactor autocomplete methods

kaustubh-nair opened the pull request:
As per the new changes, adding the footer text for typeaheads will now cause changes only in the `generic_autocomplete` method. This should decrease a lot of redundancy faced in #540

neiljp commented on zulipterminal/helper.py:
```diff
@@ -477,6 +477,14 @@ def canonicalize_color(color: str) -> str:
 
 
+def styled_typeahead(typeaheads: List[str],
+                     wrapping_text: str) -> List[str]:
```
I assume this method is a precursor to make styled (with bugdown/markdown) versions separate from the versions in the bottom bar? If so, this feels like it belongs with that work.

We could include it now, but it's clearer seeing it 'in action' - demonstrating why it's much better - though arguably it does that to some extent already.

kaustubh-nair replied:
Right, this is mainly to decrease the clutter in the footer PR. It'll be better if I put it there :+1:

neiljp commented on zulipterminal/helper.py:
```diff
@@ -442,7 +442,7 @@ def match_stream(stream: Any, text: str) -> bool:
 
 
-def match_groups(group_name: str, text: str) -> bool:
+def match_group(group_name: str, text: str) -> bool:
```
I can see why this minimal rename could be useful, but you can explain why in the commit message. This helps ensure that reviewers and those who look at commits later understand your motivation - did you just change the name to add a commit? ;)

kaustubh-nair replied:
It was really bothering me that all other `match_` methods were singular, but `match_groups` had an `s` at the end. :stuck_out_tongue_closed_eyes:

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -132,18 +130,17 @@ def autocomplete_mentions(self, text: str, prefix_string: str
-                        if match_stream(stream, text[1:])]
+
+        streams = filtered_typeahead_names(streams_list, text, match_stream, 1)
+        stream_names = [stream[0] for stream in streams]
```
This change has two list comprehensions, one inside a function, compared to just one list comprehension. That doesn't make it necessarily unwise, but if this can be seen more clearly worthwhile in the context of the other PR, let's put it there?

kaustubh-nair replied:
Done :+1:

kaustubh-nair commented on tests/ui_tools/test_boxes.py:
```diff
@@ -17,6 +17,24 @@ def write_box(self, mocker):
+        return write_box
+
+    @pytest.fixture()
+    def write_box_streams(self, mocker, streams_fixture, write_box):
```
Right, I'll put this over with the footer PR. These fixtures get used multiple times over there.

kaustubh-nair commented on tests/ui_tools/test_boxes.py:
```diff
@@ -166,6 +167,6 @@ def test_autocomplete_streams(self, write_box_streams, text,
                                  text, state, required_typeahead):
-        typeahead_string = write_box.autocomplete_emojis(text,
-                                                         state, emojis_fixture)
+        zulipterminal.emoji_names.EMOJI_NAMES = emojis_fixture
```
Yes, I wanted to do that too. But is it possible to mock local variables? I couldn't find any reasonable way to do it. Please let me know if you have any ideas

kaustubh-nair replied:
I don't think we would need an emoji version of `write_box`; `write_box` doesn't contain any emoji related attributes. Unless you're suggesting setting `emoji_names` as a `write_box` attribute?

neiljp replied:
The mock should be possible; I think this is more straightforward since `emoji_names` can be fully mocked.

kaustubh-nair replied:
Good idea :+1: done

kaustubh-nair commented:
Thanks for the review @neiljp 
I have removed the footer PR related commits , so this one becomes much simpler now. 
Commits 2 and 3 are the most important ones - they should really help avoid duplication while adding new features. I'll add the commits I removed from here to the footer PR once these two commits are merged.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -41,82 +41,70 @@ def test_init(self, write_box):
-    def test_generic_autocomplete(self, mocker, write_box, text,
-                                  prefix_string, state):
-        autocomplete_mentions = mocker.patch(WRITEBOX+'.autocomplete_mentions')
+    def test_generic_autocomplete_no_prefix(self, mocker, write_box, text,
```
You can also note that this test set is split into it's own test.

kaustubh-nair replied:
Done :+1:

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -96,18 +96,23 @@ def stream_box_view(self, caption: str='', title: str='') -> None:
+            typeahead = self.autocomplete_streams(text)
         elif text.startswith(':'):
-            return self.autocomplete_emojis(text, state)
+            typeahead = self.autocomplete_emojis(text)
```
We may be able to simplify this block further - it seems rather repetitive - though we don't have to do it in this PR necessarily.

kaustubh-nair replied:
I'll probably look at this in the footer PR, some of the changes made there could affect the refactor here.

kaustubh-nair commented:
@neiljp Thanks for the review! I've updated the PR according to your suggestions.

neiljp commented:
@kaustubh-nair I made the few small changes remaining and pushed this just now - thanks :tada:

There were a few minor typos in the function names in the commit titles/descriptions.

Other differences were just with regard to how the mocking can be simplified (no need to import the module to mock, and mock attributes can be set on initialization).
